### PR TITLE
Add Identity Access Management (IAM) to the Storage API

### PR DESCRIPTION
**Pull request by rybosome**
Adds support for bucket-level IAM (currently in limited alpha). More information
about IAM in Google Cloud Storage can be found at
https://cloud.google.com/storage/docs/access-control/iam

### CODE REVIEW

**Comment by garrettjonesgoogle** [comment]
@neozwu could you do a first pass on this PR?

File: google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java [hunk 1]
```diff
@@ -1440,4 +1445,49 @@ public void testReadCompressedBlob() throws IOException {
...
+    	    StorageRoles.legacyBucketOwner(), (Set<Identity>) newHashSet(projectOwner, projectEditor),
+    	    StorageRoles.legacyBucketReader(), newHashSet(projectViewer));
+    Map<com.google.cloud.Role, Set<Identity>> bindingsWithPublicRead =
+        ImmutableMap.of(
+    	    StorageRoles.legacyBucketOwner(), (Set<Identity>) newHashSet(projectOwner, projectEditor),
+    	    StorageRoles.legacyBucketReader(), newHashSet(projectViewer),
+          StorageRoles.legacyObjectReader(), newHashSet(Identity.allUsers()));
+
+    // Validate getting policy.
+    Policy currentPolicy = storage.getPolicy(BUCKET);
```
Comment (neozwu) [inline google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java]: This currently failed on my local machine with the following error.  Is it expected ?

com.google.cloud.storage.StorageException: Invalid argument.

	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:188)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.getPolicy(HttpStorageRpc.java:845)
	at com.google.cloud.storage.StorageImpl$35.call(StorageImpl.java:870)
	at com.google.cloud.storage.StorageImpl$35.call(StorageImpl.java:867)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:92)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:47)
	at com.google.cloud.storage.StorageImpl.getPolicy(StorageImpl.java:867)
	at com.google.cloud.storage.it.ITStorageTest.testBucketPolicy(ITStorageTest.java:1466)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:237)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
{
  "code" : 400,
  "errors" : [ {
    "domain" : "global",
    "message" : "Invalid argument.",
    "reason" : "invalid"
  } ],
  "message" : "Invalid argument."
}
	at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:145)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:113)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:40)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1.interceptResponse(AbstractGoogleClientRequest.java:321)
	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1056)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:419)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.getPolicy(HttpStorageRpc.java:843)
	... 35 more

Comment (rybosome) [inline google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java]: Yes, that's expected; the IAM API is whitelist-only at this point.

Comment (rybosome) [inline google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java]: (I should add that my test project is whitelisted and these tests did pass there, so the efficacy of the tests has been verified).

**Comment by garrettjonesgoogle** [comment]
@michaelbausor could you take a look too?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/testing/ApiPolicyComparator.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+  public static final ApiPolicyComparator INSTANCE = new ApiPolicyComparator();
+
+  private ApiPolicyComparator() {
+    // Intentionally left blank.
+  }
+
+  @Override
+  public int compare(Policy p1, Policy p2) {
+    int etagComparison = p1.getEtag().compareTo(p2.getEtag());
```
Comment (michaelbausor) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/testing/ApiPolicyComparator.java]: Do we want to handle null values? Or is an exception the right response?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/testing/ApiPolicyComparator.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    if (map1.size() != map2.size()) {
+      return map1.size() - map2.size();
+    }
+    
+    for (Map.Entry<String, Set<String>> entry : map1.entrySet()) {
+      String role = entry.getKey();
+      Set<String> p1Members = entry.getValue();
+      Set<String> p2Members = map2.get(role);
+      if (p2Members == null) {
+        return -1;
```
Comment (michaelbausor) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/testing/ApiPolicyComparator.java]: This comparator doesn't provide a total ordering on Policy objects. In the case of two different policies with the same sized bindings map, we will return -1 regardless of which one was called first.

If we only care about equality and not a total ordering, we should not use a comparator.

(See https://docs.oracle.com/javase/7/docs/api/java/util/Comparator.html)

File: google-cloud-storage/src/test/java/com/google/cloud/storage/testing/ApiPolicyComparatorTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+import com.google.api.services.storage.model.Policy;
+import com.google.api.services.storage.model.Policy.Bindings;
+import com.google.common.collect.ImmutableList;
+
+public class ApiPolicyComparatorTest {
```
Comment (michaelbausor) [inline google-cloud-storage/src/test/java/com/google/cloud/storage/testing/ApiPolicyComparatorTest.java]: If we are going to use a comparator, we should test the not-equals case, and also test that if `compare(p1,p2)==-1` then `compare(p2,p1)==1`

**Comment by rybosome** [comment]
@michaelbausor This is ready for another pass.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/testing/ApiPolicyMatcher.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+import com.google.api.services.storage.model.Policy;
+import com.google.api.services.storage.model.Policy.Bindings;
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * Matches two {@link Policy} instances, which may have lists of {@link Bindings} that are not in
+ * the same order but which are still logically equivalent.
+ */
+public class ApiPolicyMatcher implements IArgumentMatcher {
```
Comment (michaelbausor) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/testing/ApiPolicyMatcher.java]: It is necessary for this to be in `src/main`? If we put it in `src/test` then we could avoid needing to update the dependency.

**Comment by rybosome** [comment]
@michaelbausor Took your suggestion, now ready for review.

**Comment by michaelbausor** [comment]
LGTM

**Pull request merged**